In [1]:
from scenedetect import detect, ContentDetector, split_video_ffmpeg, SceneManager, AdaptiveDetector, ThresholdDetector, HashDetector, HistogramDetector, StatsManager, save_images, open_video
video_path = '/Users/spectatr/Downloads/soccernet_data/england_epl/2014-2015/2015-02-21 - 18-00 Chelsea 1 - 1 Burnley/2_720p.mkv'
scene_list = detect(video_path, ContentDetector()) #thresholod inside content detection: Content got 249 frames.
""" In above, we can add show_progress=True"""
print(scene_list)
split_video_ffmpeg(video_path, scene_list, output_file_template='$VIDEO_NAME-Scene-$SCENE_NUMBER.mp4')
save_images(scene_list, open_video(video_path), image_extension='.jpg', image_name_template='$VIDEO_NAME-Scene-$SCENE_NUMBER-$IMAGE_NUMBER', output_dir='./save_images')

[(00:00:00.000 [frame=0, fps=25.000], 00:00:33.720 [frame=843, fps=25.000]),
 (00:00:33.720 [frame=843, fps=25.000], 00:00:37.600 [frame=940, fps=25.000]),
 (00:00:37.600 [frame=940, fps=25.000], 00:00:43.680 [frame=1092, fps=25.000]),
 (00:00:43.680 [frame=1092, fps=25.000],
  00:01:06.720 [frame=1668, fps=25.000]),
 (00:01:06.720 [frame=1668, fps=25.000],
  00:01:13.680 [frame=1842, fps=25.000]),
 (00:01:13.680 [frame=1842, fps=25.000],
  00:01:43.520 [frame=2588, fps=25.000]),
 (00:01:43.520 [frame=2588, fps=25.000],
  00:01:44.840 [frame=2621, fps=25.000]),
 (00:01:44.840 [frame=2621, fps=25.000],
  00:01:48.960 [frame=2724, fps=25.000]),
 (00:01:48.960 [frame=2724, fps=25.000],
  00:02:00.840 [frame=3021, fps=25.000]),
 (00:02:00.840 [frame=3021, fps=25.000],
  00:02:05.640 [frame=3141, fps=25.000]),
 (00:02:05.640 [frame=3141, fps=25.000],
  00:02:10.160 [frame=3254, fps=25.000]),
 (00:02:10.160 [frame=3254, fps=25.000],
  00:02:16.640 [frame=3416, fps=25.000]),
 (00:02:16.640 [f

In [6]:
from scenedetect import detect, ContentDetector, split_video_ffmpeg, SceneManager, AdaptiveDetector, ThresholdDetector, HashDetector, HistogramDetector, StatsManager
video_path = '/Users/spectatr/Downloads/soccernet_data/england_epl/2014-2015/2015-02-21 - 18-00 Chelsea 1 - 1 Burnley/2_720p.mkv'
scene_list = detect(video_path, HistogramDetector()) #thresholod inside content detection
scene_list
print(f"Lenght of scene list is {len(scene_list)}") 

Lenght of scene list is 318


# 
- custom our algorithm took 20 minutes to extract framechange with 209 frames.
- Lenght is 307 for HashDetector took 2minutes.
- Scene Detector is 153 frames in 30 seconds
- content dectector took 2 minutes for 270 frames
- Adaptive only created 152 = wrong
- Histogram detector took = 2 minuste with 318

In [2]:
for i, scene in enumerate(scene_list):
    print('    Scene %2d: Start %s / Frame %d, End %s / Frame %d' % (
        i+1,
        scene[0].get_timecode(), scene[0].get_frames(),
        scene[1].get_timecode(), scene[1].get_frames(),))


    Scene  1: Start 00:00:00.000 / Frame 0, End 00:00:33.720 / Frame 843
    Scene  2: Start 00:00:33.720 / Frame 843, End 00:00:37.600 / Frame 940
    Scene  3: Start 00:00:37.600 / Frame 940, End 00:00:43.680 / Frame 1092
    Scene  4: Start 00:00:43.680 / Frame 1092, End 00:01:06.720 / Frame 1668
    Scene  5: Start 00:01:06.720 / Frame 1668, End 00:01:13.680 / Frame 1842
    Scene  6: Start 00:01:13.680 / Frame 1842, End 00:01:43.520 / Frame 2588
    Scene  7: Start 00:01:43.520 / Frame 2588, End 00:01:44.840 / Frame 2621
    Scene  8: Start 00:01:44.840 / Frame 2621, End 00:01:48.960 / Frame 2724
    Scene  9: Start 00:01:48.960 / Frame 2724, End 00:02:00.840 / Frame 3021
    Scene 10: Start 00:02:00.840 / Frame 3021, End 00:02:05.640 / Frame 3141
    Scene 11: Start 00:02:05.640 / Frame 3141, End 00:02:10.160 / Frame 3254
    Scene 12: Start 00:02:10.160 / Frame 3254, End 00:02:16.640 / Frame 3416
    Scene 13: Start 00:02:16.640 / Frame 3416, End 00:02:51.600 / Frame 4290
    Sc

In [ ]:
split_video_ffmpeg(video_path, scene_list)

In [7]:
#save-images command:save an image of the first and last frame of each detected scene via the
# list-scense, split-video

In [14]:
from scenedetect import detect, ContentDetector, split_video_ffmpeg, SceneManager, AdaptiveDetector, ThresholdDetector, HashDetector, HistogramDetector, StatsManager, open_video
video_path = '/Users/spectatr/Downloads/soccernet_data/england_epl/2014-2015/2015-02-21 - 18-00 Chelsea 1 - 1 Burnley/2_720p.mkv'
video = open_video(video_path)
scene_manager = SceneManager(stats_manager=StatsManager())
# scene_manager.add_detector(ThresholdDetector()) # Can add multiple detectors. Threshold is to detect fades in/out
scene_manager.add_detector(AdaptiveDetector()) 
scene_manager.detect_scenes(video=video)
scene_list = scene_manager.get_scene_list()
print(scene_list, scene_manager.get_cut_list(), scene_manager.get_event_list())
# print_scenes(scene_list=scene_list)
scene_manager.stats_manager.save_to_csv(csv_file='/Users/spectatr/Downloads/general_scripts/model_improvement_experiments/stats.csv')

In [21]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.widgets import Slider
from scipy import signal
import warnings
import os

warnings.filterwarnings('ignore')

# Set up plotting style
plt.style.use('seaborn-v0_8')
sns.set_palette("husl")

def load_and_analyze_data(csv_file_path):
    """Load CSV data and perform initial analysis"""
    # Load the data
    df = pd.read_csv(csv_file_path)
    
    # Convert timecode to seconds for easier analysis
    def timecode_to_seconds(timecode):
        parts = timecode.split(':')
        hours, minutes, seconds = float(parts[0]), float(parts[1]), float(parts[2])
        return hours * 3600 + minutes * 60 + seconds
    
    df['time_seconds'] = df['Timecode'].apply(timecode_to_seconds)
    
    return df

def create_comprehensive_plots(df):
    """Create multiple visualization plots for threshold analysis"""
    
    # Create figure with subplots
    fig = plt.figure(figsize=(20, 15))
    
    # 1. Time series plot of all metrics
    ax1 = plt.subplot(3, 2, 1)
    metrics = ['content_val', 'delta_edges', 'delta_hue', 'delta_lum', 'delta_sat']
    for metric in metrics:
        plt.plot(df['time_seconds'], df[metric], label=metric, linewidth=2, alpha=0.8)
    plt.xlabel('Time (seconds)')
    plt.ylabel('Detection Values')
    plt.title('All Detection Metrics Over Time')
    plt.legend()
    plt.grid(True, alpha=0.3)
    
    # 2. Focus on content_val with potential thresholds
    ax2 = plt.subplot(3, 2, 2)
    plt.plot(df['time_seconds'], df['content_val'], 'b-', linewidth=2, label='content_val')
    
    # Add threshold lines
    content_mean = df['content_val'].mean()
    content_std = df['content_val'].std()
    thresholds = [
        content_mean + content_std,
        content_mean + 2*content_std,
        content_mean + 1.5*content_std
    ]
    
    colors = ['red', 'orange', 'green']
    labels = ['Mean + 1σ', 'Mean + 2σ', 'Mean + 1.5σ']
    
    for threshold, color, label in zip(thresholds, colors, labels):
        plt.axhline(y=threshold, color=color, linestyle='--', alpha=0.7, label=f'{label}: {threshold:.2f}')
    
    plt.xlabel('Time (seconds)')
    plt.ylabel('Content Value')
    plt.title('Content Value with Suggested Thresholds')
    plt.legend()
    plt.grid(True, alpha=0.3)
    
    # 3. Distribution analysis
    ax3 = plt.subplot(3, 2, 3)
    plt.hist(df['content_val'], bins=30, alpha=0.7, edgecolor='black')
    for threshold, color, label in zip(thresholds, colors, labels):
        plt.axvline(x=threshold, color=color, linestyle='--', linewidth=2, label=f'{label}: {threshold:.2f}')
    plt.xlabel('Content Value')
    plt.ylabel('Frequency')
    plt.title('Distribution of Content Values')
    plt.legend()
    plt.grid(True, alpha=0.3)
    
    # 4. Correlation heatmap
    ax4 = plt.subplot(3, 2, 4)
    correlation_matrix = df[metrics].corr()
    sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', center=0, 
                square=True, fmt='.2f', cbar_kws={"shrink": .8})
    plt.title('Correlation Between Detection Metrics')
    
    # 5. Box plot comparison
    ax5 = plt.subplot(3, 2, 5)
    df_melted = pd.melt(df, id_vars=['Frame Number'], value_vars=metrics, 
                       var_name='Metric', value_name='Value')
    sns.boxplot(data=df_melted, x='Metric', y='Value')
    plt.xticks(rotation=45)
    plt.title('Distribution Comparison of All Metrics')
    plt.tight_layout()
    
    # 6. Derivative analysis for content_val
    ax6 = plt.subplot(3, 2, 6)
    content_derivative = np.gradient(df['content_val'])
    plt.plot(df['time_seconds'], content_derivative, 'purple', linewidth=2, label='Content Value Derivative')
    plt.axhline(y=0, color='black', linestyle='-', alpha=0.3)
    
    # Mark significant changes
    significant_changes = np.abs(content_derivative) > np.std(content_derivative) * 2
    plt.scatter(df['time_seconds'][significant_changes], 
               content_derivative[significant_changes], 
               color='red', s=50, alpha=0.7, label='Significant Changes')
    
    plt.xlabel('Time (seconds)')
    plt.ylabel('Rate of Change')
    plt.title('Content Value Rate of Change (Potential Scene Cuts)')
    plt.legend()
    plt.grid(True, alpha=0.3)
    
    plt.tight_layout()
    return fig

def interactive_threshold_explorer(df):
    """Create an interactive plot to explore different threshold values"""
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(15, 10))
    
    # Initial threshold
    initial_threshold = df['content_val'].mean() + df['content_val'].std()
    
    # Plot 1: Time series with threshold
    line1, = ax1.plot(df['time_seconds'], df['content_val'], 'b-', linewidth=2, label='content_val')
    threshold_line, = ax1.plot(df['time_seconds'], [initial_threshold] * len(df), 
                              'r--', linewidth=2, label=f'Threshold: {initial_threshold:.2f}')
    
    # Mark detected scenes
    scenes = df['content_val'] > initial_threshold
    scene_points = ax1.scatter(df['time_seconds'][scenes], df['content_val'][scenes], 
                              color='red', s=50, alpha=0.7, label='Detected Scenes')
    
    ax1.set_xlabel('Time (seconds)')
    ax1.set_ylabel('Content Value')
    ax1.set_title('Interactive Threshold Explorer')
    ax1.legend()
    ax1.grid(True, alpha=0.3)
    
    # Plot 2: Statistics
    ax2.text(0.1, 0.8, f'Total Frames: {len(df)}', transform=ax2.transAxes, fontsize=12)
    ax2.text(0.1, 0.7, f'Detected Scenes: {np.sum(scenes)}', transform=ax2.transAxes, fontsize=12)
    ax2.text(0.1, 0.6, f'Detection Rate: {np.sum(scenes)/len(df)*100:.1f}%', transform=ax2.transAxes, fontsize=12)
    ax2.text(0.1, 0.5, f'Avg Time Between Scenes: {len(df)/np.sum(scenes)*0.04:.2f}s', 
             transform=ax2.transAxes, fontsize=12)
    ax2.set_xlim(0, 1)
    ax2.set_ylim(0, 1)
    ax2.axis('off')
    
    # Add slider
    ax_slider = plt.axes([0.2, 0.02, 0.6, 0.03])
    slider = Slider(ax_slider, 'Threshold', 
                   df['content_val'].min(), df['content_val'].max(), 
                   valinit=initial_threshold, valstep=0.1)
    
    def update_threshold(val):
        threshold = slider.val
        threshold_line.set_ydata([threshold] * len(df))
        
        # Update scene detection
        scenes = df['content_val'] > threshold
        scene_points.set_offsets(np.column_stack([df['time_seconds'][scenes], df['content_val'][scenes]]))
        
        # Update statistics
        ax2.clear()
        ax2.text(0.1, 0.8, f'Total Frames: {len(df)}', transform=ax2.transAxes, fontsize=12)
        ax2.text(0.1, 0.7, f'Detected Scenes: {np.sum(scenes)}', transform=ax2.transAxes, fontsize=12)
        ax2.text(0.1, 0.6, f'Detection Rate: {np.sum(scenes)/len(df)*100:.1f}%', transform=ax2.transAxes, fontsize=12)
        if np.sum(scenes) > 0:
            ax2.text(0.1, 0.5, f'Avg Time Between Scenes: {len(df)/np.sum(scenes)*0.04:.2f}s', 
                     transform=ax2.transAxes, fontsize=12)
        ax2.text(0.1, 0.4, f'Current Threshold: {threshold:.2f}', transform=ax2.transAxes, fontsize=12)
        ax2.set_xlim(0, 1)
        ax2.set_ylim(0, 1)
        ax2.axis('off')
        
        threshold_line.set_label(f'Threshold: {threshold:.2f}')
        ax1.legend()
        fig.canvas.draw()
    
    slider.on_changed(update_threshold)
    
    plt.tight_layout()
    return fig, slider

def generate_recommendations(df):
    """Generate threshold recommendations based on statistical analysis"""
    content_vals = df['content_val']
    
    # Statistical measures
    mean_val = content_vals.mean()
    std_val = content_vals.std()
    median_val = content_vals.median()
    q75 = content_vals.quantile(0.75)
    q90 = content_vals.quantile(0.90)
    q95 = content_vals.quantile(0.95)
    
    print("=== THRESHOLD RECOMMENDATIONS ===\n")
    
    print(f"Statistical Summary:")
    print(f"  Mean: {mean_val:.2f}")
    print(f"  Std Dev: {std_val:.2f}")
    print(f"  Median: {median_val:.2f}")
    print(f"  75th Percentile: {q75:.2f}")
    print(f"  90th Percentile: {q90:.2f}")
    print(f"  95th Percentile: {q95:.2f}\n")
    
    recommendations = [
        ("Conservative (fewer scenes)", q95, "Detects only the most significant changes"),
        ("Moderate", mean_val + 1.5*std_val, "Balanced approach, good starting point"),
        ("Sensitive", q75, "Detects more subtle changes"),
        ("Very Sensitive", median_val, "May catch too many minor changes")
    ]
    
    print("Recommended Thresholds:")
    for name, threshold, description in recommendations:
        scenes_detected = np.sum(content_vals > threshold)
        detection_rate = scenes_detected / len(df) * 100
        print(f"  {name}: {threshold:.2f}")
        print(f"    - {description}")
        print(f"    - Would detect {scenes_detected} scenes ({detection_rate:.1f}% of frames)")
        if scenes_detected > 0:
            avg_time_between = len(df) / scenes_detected * 0.04  # assuming 25fps
            print(f"    - Average time between scenes: {avg_time_between:.2f}s")
        print()

def create_all_plots(df):
    """Create comprehensive analysis plots"""
    # Calculate thresholds
    mean_val = df['content_val'].mean()
    std_val = df['content_val'].std()
    
    thresholds = {
        'Conservative': df['content_val'].quantile(0.95),
        'Moderate': mean_val + 1.5*std_val,
        'Sensitive': df['content_val'].quantile(0.75)
    }
    
    # Create figure with subplots
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    fig.suptitle('PySceneDetect Threshold Analysis', fontsize=16, fontweight='bold')
    
    # 1. Time series with thresholds
    ax = axes[0, 0]
    ax.plot(df['time_seconds'], df['content_val'], 'b-', linewidth=2, label='content_val')
    colors = ['red', 'orange', 'green']
    for (name, threshold), color in zip(thresholds.items(), colors):
        ax.axhline(y=threshold, color=color, linestyle='--', alpha=0.7, label=f'{name}: {threshold:.1f}')
        scenes = df['content_val'] > threshold
        ax.scatter(df['time_seconds'][scenes], df['content_val'][scenes], color=color, s=20, alpha=0.6)
    ax.set_xlabel('Time (seconds)')
    ax.set_ylabel('Content Value')
    ax.set_title('Content Value with Thresholds')
    ax.legend()
    ax.grid(True, alpha=0.3)
    
    # 2. Distribution
    ax = axes[0, 1]
    ax.hist(df['content_val'], bins=20, alpha=0.7, edgecolor='black')
    for (name, threshold), color in zip(thresholds.items(), colors):
        ax.axvline(x=threshold, color=color, linestyle='--', linewidth=2)
    ax.set_xlabel('Content Value')
    ax.set_ylabel('Frequency')
    ax.set_title('Value Distribution')
    ax.grid(True, alpha=0.3)
    
    # 3. All metrics comparison
    ax = axes[0, 2]
    metrics = ['content_val', 'delta_edges', 'delta_hue', 'delta_lum', 'delta_sat']
    for i, metric in enumerate(metrics):
        ax.plot(df['time_seconds'], df[metric], label=metric, alpha=0.8)
    ax.set_xlabel('Time (seconds)')
    ax.set_ylabel('Values')
    ax.set_title('All Detection Metrics')
    ax.legend()
    ax.grid(True, alpha=0.3)
    
    # 4. Threshold comparison
    ax = axes[1, 0]
    threshold_values = np.linspace(df['content_val'].min(), df['content_val'].max(), 30)
    scene_counts = [np.sum(df['content_val'] > t) for t in threshold_values]
    detection_rates = [count/len(df)*100 for count in scene_counts]
    
    ax.plot(threshold_values, detection_rates, 'b-', linewidth=2)
    for (name, threshold), color in zip(thresholds.items(), colors):
        ax.axvline(x=threshold, color=color, linestyle='--', alpha=0.7)
    ax.set_xlabel('Threshold Value')
    ax.set_ylabel('Detection Rate (%)')
    ax.set_title('Threshold vs Detection Rate')
    ax.grid(True, alpha=0.3)
    
    # 5. Scene timing (using moderate threshold)
    ax = axes[1, 1]
    moderate_threshold = thresholds['Moderate']
    scenes = df['content_val'] > moderate_threshold
    if np.sum(scenes) > 1:
        scene_indices = np.where(scenes)[0]
        scene_gaps = np.diff(scene_indices) * 0.04  # Convert to seconds
        ax.hist(scene_gaps, bins=min(10, len(scene_gaps)), alpha=0.7, edgecolor='black')
        ax.axvline(np.mean(scene_gaps), color='red', linestyle='--', 
                  label=f'Mean: {np.mean(scene_gaps):.1f}s')
        ax.set_xlabel('Time Between Scenes (s)')
        ax.set_ylabel('Frequency')
        ax.set_title('Scene Gap Distribution')
        ax.legend()
    else:
        ax.text(0.5, 0.5, 'Not enough scenes\nfor timing analysis', 
               ha='center', va='center', transform=ax.transAxes)
        ax.set_title('Scene Timing')
    ax.grid(True, alpha=0.3)
    
    # 6. Recommendations table
    ax = axes[1, 2]
    ax.axis('off')
    table_data = []
    for name, threshold in thresholds.items():
        scenes = np.sum(df['content_val'] > threshold)
        rate = scenes / len(df) * 100
        avg_time = f"{len(df) / scenes * 0.04:.1f}s" if scenes > 0 else "N/A"
        table_data.append([name, f'{threshold:.1f}', f'{scenes}', f'{rate:.0f}%', avg_time])
    
    table = ax.table(cellText=table_data,
                    colLabels=['Type', 'Threshold', 'Scenes', 'Rate', 'Avg Gap'],
                    cellLoc='center', loc='center')
    table.auto_set_font_size(False)
    table.set_fontsize(10)
    table.scale(1, 1.5)
    ax.set_title('Recommendations', pad=20)
    
    plt.tight_layout()
    return fig, thresholds

def save_analysis(df, output_name="scene_analysis"):
    """Save complete analysis to PNG files and generate recommendations"""
    # Create output directory if it doesn't exist
    output_dir = f"{output_name}_images"
    os.makedirs(output_dir, exist_ok=True)
    
    # Create main analysis plot
    fig1, thresholds = create_all_plots(df)
    main_plot_path = os.path.join(output_dir, f"{output_name}_main_analysis.png")
    fig1.savefig(main_plot_path, dpi=300, bbox_inches='tight', facecolor='white')
    plt.close(fig1)
    
    # Create comprehensive plots
    fig2 = create_comprehensive_plots(df)
    comprehensive_plot_path = os.path.join(output_dir, f"{output_name}_comprehensive_analysis.png")
    fig2.savefig(comprehensive_plot_path, dpi=300, bbox_inches='tight', facecolor='white')
    plt.close(fig2)
    
    # Create individual plots for better readability
    create_individual_plots(df, output_dir, output_name)
    
    print(f"✅ Analysis saved to PNG files:")
    print(f"   📁 Directory: {output_dir}/")
    print(f"   📊 {output_name}_main_analysis.png")
    print(f"   📊 {output_name}_comprehensive_analysis.png")
    print(f"   📊 Individual plot files")
    
    return thresholds

def create_individual_plots(df, output_dir, output_name):
    """Create individual plots for better readability"""
    
    # 1. Time series plot
    fig, ax = plt.subplots(figsize=(12, 6))
    ax.plot(df['time_seconds'], df['content_val'], 'b-', linewidth=2, label='content_val')
    
    # Add threshold lines
    mean_val = df['content_val'].mean()
    std_val = df['content_val'].std()
    thresholds = {
        'Conservative': df['content_val'].quantile(0.95),
        'Moderate': mean_val + 1.5*std_val,
        'Sensitive': df['content_val'].quantile(0.75)
    }
    
    colors = ['red', 'orange', 'green']
    for (name, threshold), color in zip(thresholds.items(), colors):
        ax.axhline(y=threshold, color=color, linestyle='--', alpha=0.7, label=f'{name}: {threshold:.1f}')
        scenes = df['content_val'] > threshold
        ax.scatter(df['time_seconds'][scenes], df['content_val'][scenes], color=color, s=30, alpha=0.6)
    
    ax.set_xlabel('Time (seconds)')
    ax.set_ylabel('Content Value')
    ax.set_title('Content Value Over Time with Recommended Thresholds')
    ax.legend()
    ax.grid(True, alpha=0.3)
    
    plt.tight_layout()
    fig.savefig(os.path.join(output_dir, f"{output_name}_timeseries.png"), 
                dpi=300, bbox_inches='tight', facecolor='white')
    plt.close(fig)
    
    # 2. Distribution plot
    fig, ax = plt.subplots(figsize=(10, 6))
    ax.hist(df['content_val'], bins=30, alpha=0.7, edgecolor='black', color='skyblue')
    
    for (name, threshold), color in zip(thresholds.items(), colors):
        ax.axvline(x=threshold, color=color, linestyle='--', linewidth=2, label=f'{name}: {threshold:.1f}')
    
    ax.set_xlabel('Content Value')
    ax.set_ylabel('Frequency')
    ax.set_title('Distribution of Content Values with Thresholds')
    ax.legend()
    ax.grid(True, alpha=0.3)
    
    plt.tight_layout()
    fig.savefig(os.path.join(output_dir, f"{output_name}_distribution.png"), 
                dpi=300, bbox_inches='tight', facecolor='white')
    plt.close(fig)
    
    # 3. All metrics comparison
    fig, ax = plt.subplots(figsize=(14, 8))
    metrics = ['content_val', 'delta_edges', 'delta_hue', 'delta_lum', 'delta_sat']
    
    for metric in metrics:
        ax.plot(df['time_seconds'], df[metric], label=metric, linewidth=2, alpha=0.8)
    
    ax.set_xlabel('Time (seconds)')
    ax.set_ylabel('Detection Values')
    ax.set_title('All Detection Metrics Over Time')
    ax.legend()
    ax.grid(True, alpha=0.3)
    
    plt.tight_layout()
    fig.savefig(os.path.join(output_dir, f"{output_name}_all_metrics.png"), 
                dpi=300, bbox_inches='tight', facecolor='white')
    plt.close(fig)

# Example usage:
if __name__ == "__main__":
    # Replace with your CSV file path
    csv_file = "/Users/spectatr/Downloads/general_scripts/model_improvement_experiments/stats.csv"
    # df = pd.DataFrame(sample_data)
    df = pd.read_csv(csv_file, header=0)
    df['time_seconds'] = df['Frame Number'] * 0.04  # Assuming 25fps

    # Save analysis as PNG files
    save_analysis(df, "football_analysis")
    
    # Create visualizations
    fig1 = create_comprehensive_plots(df)
    plt.show()
    
    # Create interactive explorer
    fig2, slider = interactive_threshold_explorer(df)
    plt.show()
    
    # Generate recommendations
    generate_recommendations(df)
    
    print("\n=== USAGE INSTRUCTIONS ===")
    print("1. Replace the sample data with your actual CSV file")
    print("2. Use the comprehensive plots to understand your data patterns")
    print("3. Use the interactive explorer to test different thresholds")
    print("4. Consider the recommendations based on your specific needs")
    print("5. For football matches, start with the 'Moderate' recommendation")
    print("6. Adjust based on whether you're getting too many or too few scene cuts")
    print("7. PNG files are saved in a dedicated folder for easy access")

✅ Analysis saved to PNG files:
   📁 Directory: football_analysis_images/
   📊 football_analysis_main_analysis.png
   📊 football_analysis_comprehensive_analysis.png
   📊 Individual plot files
=== THRESHOLD RECOMMENDATIONS ===

Statistical Summary:
  Mean: 8.69
  Std Dev: 6.02
  Median: 7.96
  75th Percentile: 10.06
  90th Percentile: 13.88
  95th Percentile: 19.05

Recommended Thresholds:
  Conservative (fewer scenes): 19.05
    - Detects only the most significant changes
    - Would detect 3375 scenes (5.0% of frames)
    - Average time between scenes: 0.80s

  Moderate: 17.72
    - Balanced approach, good starting point
    - Would detect 4069 scenes (6.0% of frames)
    - Average time between scenes: 0.66s

  Sensitive: 10.06
    - Detects more subtle changes
    - Would detect 16875 scenes (25.0% of frames)
    - Average time between scenes: 0.16s

  Very Sensitive: 7.96
    - May catch too many minor changes
    - Would detect 33749 scenes (50.0% of frames)
    - Average time betwe

In [ ]:
# Live streaming
def live_video_processing():
    import cv2
    
    # Setup for webcam
    cap = cv2.VideoCapture(0)
    video = VideoCaptureAdapter(cap)
    
    scene_manager = SceneManager()
    scene_manager.add_detector(ContentDetector())
    
    # Process with callback for real-time detection
    def scene_callback(frame_num, start_time, end_time):
        print(f"Scene detected: {start_time} to {end_time}")
    
    # Detect scenes with callback
    scene_manager.detect_scenes(
        video=video,
        duration=1000,  # Process 1000 frames
        callback=scene_callback
    )